### PR TITLE
[1/2] Registry interface for custom quantization functional backend

### PR DESCRIPTION
**Pull request by realAsma**
## What does this PR do?

**Type of change:** ? new feature

**Overview:** 

Add support to enable custom emulated quantization backend. See :meth:`register_quant_backend <modelopt.torch.quantization.nn.modules.tensor_quantizer.register_quant_backend>`` for more details. 

## Usage

See an example in ``tests/unit/torch/quantization/test_custom_backend.py``.

## Testing

See the unites

## Before your PR is "*Ready for review*"

- **Make sure you read and follow [Contributor guidelines](https://github.com/NVIDIA/TensorRT-Model-Optimizer/blob/main/CONTRIBUTING.md)** and your commits are signed.
- **Is this change backward compatible?**: Yes/No 
- **Did you write any new necessary tests?**: Yes/No
- **Did you add or update any necessary documentation?**: Yes/No
- **Did you update [Changelog](https://github.com/NVIDIA/TensorRT-Model-Optimizer/blob/main/CHANGELOG.rst)?**: Yes/No 

## Summary by CodeRabbit

* **New Features**
  * Support for pluggable custom quantization backends with configurable backend args.
  * Quantization config accepts string-based bit specifications.

* **Improvements**
  * Algorithm parameter made optional in quantization calls.
  * Device-aware CUDA handling to ensure correct device placement during quantization.
  * Optional Triton-accelerated FP4 path for improved performance where available.
  * Warning and automatic disabling when gradient_accumulation_fusion is incompatible.

* **Tests**
  * Added unit test covering custom backend usage; updated GPU tests for non-current device scenarios.

### CODE REVIEW

File: modelopt/torch/quantization/plugins/megatron.py [hunk 1]
```diff
@@ -231,6 +231,14 @@ def _setup(self):
...
             mcore_parallel.get_tensor_model_parallel_group(),
         )
+
+        if getattr(self, "gradient_accumulation_fusion", False):
+            warnings.warn(
+                "gradient_accumulation_fusion is not supported with ModelOpt quantization. "
+                "Setting gradient_accumulation_fusion to False."
+            )
+            self.gradient_accumulation_fusion = False
+
```
Comment (realAsma) [inline modelopt/torch/quantization/plugins/megatron.py:241]: @jenchen13  @ChenhanYu this should fix the occasional errors people run into by unknowingly enabling `gradient_accumulation_fusion`

File: CHANGELOG.rst [hunk 2]
```diff
@@ -8,9 +8,10 @@ Model Optimizer Changelog (Linux)
 
 **New Features**
 
-- Add flag ``op_types_to_exclude_fp16`` in ONNX quantization to exclude ops from being converted to FP16/BF16. Alternatively, for custom TensorRT ops, this can also be done by indicating ``'fp32'`` precision in ``trt_plugins_precision``.
 - Add LoRA mode support for MCore in a new peft submodule: ``modelopt.torch.peft.update_model(model, LORA_CFG)``.
 - Support PTQ and fakequant in vLLM for fast evaluation of arbitrary quantization formats. See ``examples/vllm_serve`` for more details.
+- Add support to enable custom emulated quantization backend. See :meth:`register_quant_backend <modelopt.torch.quantization.nn.modules.tensor_quantizer.register_quant_backend>`` for more details. See an example in ``tests/unit/torch/quantization/test_custom_backend.py``.
+- Add flag ``op_types_to_exclude_fp16`` in ONNX quantization to exclude ops from being converted to FP16/BF16. Alternatively, for custom TensorRT ops, this can also be done by indicating ``'fp32'`` precision in ``trt_plugins_precision``.
```
Comment (cjluo-nv) [inline CHANGELOG.rst:14]: @ajrasane could you review this section?

File: modelopt/torch/quantization/nn/modules/tensor_quantizer.py [hunk 3]
```diff
@@ -56,10 +57,58 @@
...
+
+    Args:
+        name: The name of the backend.
+        entrypoint: The entrypoint of the backend. The entrypoint should be a callable that takes in
+            the inputs and the tensor quantizer as arguments and returns the quantized tensor.
+            See :class:`modelopt.torch.quantization.config.QuantizerAttributeConfig`
+            for details on choosing from the registered backends via the ``backend`` and
+            ``backend_extra_args`` fields.
+    """
+    if not isinstance(name, str) or not name:
```
Comment (jenchen13) [inline modelopt/torch/quantization/nn/modules/tensor_quantizer.py:85]: is it necessary to check repeatedly if name is instance str? where do we expect possible non-str input?

Comment (jenchen13) [reply modelopt/torch/quantization/nn/modules/tensor_quantizer.py:85]: also would it make more sense for `register_quant_backend` to be a method in the TensorQuantizer since you are registering the quant backend for a specific instance of a TensorQuantizer? instead of having a callable which takes in a tensor and the quantizer instance

Currently when you register a backend for a specific quantizer it modifies the global variable `_QUANT_FUNCTIONAL_BACKENDS` whereas from the code it seems the backend is specific to the scope of an individual instance of the quantizer. It also seems odd to keep overwriting the backend if you have multiple quantizers that use the same custom backend.

File: modelopt/torch/quantization/plugins/megatron.py (hunk 1 above)
Comment (jenchen13) [reply modelopt/torch/quantization/plugins/megatron.py:241]: Thanks, why does gradient accumulation fusion not work with ModelOpt?

File: modelopt/torch/quantization/config.py [hunk 4]
```diff
@@ -675,7 +675,9 @@ class QuantizerAttributeConfig(ModeloptBaseConfig):
 
         #. Constant integer tuple (E,M) for floating point quantization emulating
             Nvidia's FPx quantization. E is the number of exponent bits and M is the number
-            of mantissa bits. Supported FPx quantization formats: FP8 (E4M3, E5M2), FP6(E3M2, E2M3), FP4(E2M1).""",
+            of mantissa bits. Supported FPx quantization formats: FP8 (E4M3, E5M2), FP6(E3M2, E2M3), FP4(E2M1).
+
+        #. String specifying the quantization format. This is current used only for custom backends.""",
```
Comment (jenchen13) [inline modelopt/torch/quantization/config.py:680]: what are the cases for custom backend where an int or tuple is not sufficient to describe the bit structure?

File: modelopt/torch/quantization/config.py [hunk 5]
```diff
@@ -707,10 +709,16 @@ def _validate_recursive(value):
...
+        if self.backend is not None:
+            # For custom backends, we don't need to validate num_bits
+            return self
+
         num_bits = self.num_bits
 
         if isinstance(num_bits, int) and num_bits < 1:
-            raise ValueError("num_bits must be a positive integer or a tuple of positive integers.")
+            raise ValueError(
+                f"num_bits must be a positive integer or a tuple of positive integers. {num_bits}"
```
Comment (jenchen13) [inline modelopt/torch/quantization/config.py:720]:
```suggestion
                f"num_bits must be a positive integer or a tuple of positive integers. Currently num_bits={num_bits}"
```

File: modelopt/torch/quantization/model_quant.py [hunk 6]
```diff
@@ -228,7 +228,7 @@ def forward_loop(model) -> None:
     Returns: A pytorch model which has been quantized and calibrated.
     """
     model = apply_mode(model, mode=[("quantize", config)], registry=QuantizeModeRegistry)
-    return calibrate(model, config["algorithm"], forward_loop=forward_loop)
+    return calibrate(model, config.get("algorithm"), forward_loop=forward_loop)
```
[the remark this answers is not in the document]
Comment (jenchen13) [reply modelopt/torch/quantization/model_quant.py:232]: +1

File: tests/unit/torch/quantization/test_custom_backend.py [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+
+import torch
+
+import modelopt.torch.quantization as mtq
+from modelopt.torch.quantization.nn import register_quant_backend, unregister_quant_backend
+
+
+def test_custom_backend_via_quantize():
+    # Define and register a simple dummy backend that adds a constant to inputs
+    def dummy_backend(inputs: torch.Tensor, tq) -> torch.Tensor:
```
Comment (jenchen13) [inline tests/unit/torch/quantization/test_custom_backend.py:19]: nit: also add TensorQuantizer type for tq

**Comment by realAsma** [comment]
Sorry this is my old PR. Let me close this. I will address the comments.

File: modelopt/torch/quantization/plugins/megatron.py (hunk 1 above)
Comment (realAsma) [reply modelopt/torch/quantization/plugins/megatron.py:241]: gradient_accumulation_fusion  fuses the gemm gradient computation and gradient accumulation to weights together within the gemm if I understand correctly. With fake quantization, the input to gemm is not the original nn.Parameter weight, but a regular tensor which is weight_quantizer(weight)

File: modelopt/torch/quantization/nn/modules/tensor_quantizer.py (hunk 3 above)
Comment (realAsma) [reply modelopt/torch/quantization/nn/modules/tensor_quantizer.py:85]: > since you are registering the quant backend for a specific instance of a TensorQuantizer
This is not true. We are registering the quant backend. Depending on the backend arg corresponding to the instance of TensorQuantizer we dispatch the corresponding from the registered backends.

File: modelopt/torch/quantization/config.py (hunk 4 above)
Comment (realAsma) [reply modelopt/torch/quantization/config.py:680]: LUTs and vector quantization etc. there is an effective bits for sure, but there is no fixed format 1:1 map. `num_bits` is just a name IMO - the value of it is used only for integer quantization - for other cases we simply use that for print.
